File: src/main/cpp/plazma/lib/task/task_helper.cpp
#include <string>

#include "task_helper.h"
#include "plazma/lib/sys/syslib.h"
#include "plazma/lib/str/strlib.h"
#include "plazma/lib/fmt/fmtlib.h"
#include "plazma/lib/fs/fslib.h"

const std::string MESSAGE_PROCESSING = "Processing...";
const std::string MESSAGE_START_TASK = "Start task: '%s'";
const std::string MESSAGE_TOTAL_TIME = "Total time: %s";

const std::string MESSAGE_NOT_SETTING = "<not setting>";

/*

void printParameter(const std::string& name, const std::string& value) {
    printParameter(name, value, 0, true);
}

void printParameter(const std::string& name, const std::string& value, int len) {
    printParameter(name, value, len, true);
}

void printParameter(const std::string& name, const std::string& value, bool setting) {
    printParameter(name, value, 0, setting);
}
*/

void printParameter(Parameter* parameter, int len) {
    if (parameter == nullptr) {
            return;            
    }
    std::string name = parameter->getName();
    std::string value = parameter->getValue();
    std::string dataType = parameter->getDataType();
    bool setting = parameter->hasValue();

    std::string displayName = formatParameterName(name, len);
    std::string displayValue = formatParameterValue(value, setting);
        
        //if ("Date".equalsIgnoreCase(dataType)) {
        //    displayValue = format
        //}        

    syslib::println(" " + displayName + ": " + displayValue);
}

// string
void printParameter(const std::string& name, const std::string& value, int len, bool setting) {
    syslib::println(" " + formatParameterName(name, len)  + ": " + formatParameterValue(value, setting));    
}

/*
// int
void printParameter(const std::string& name, const int value, int len, bool setting) {
    printParameter(name, std::to_string(value), len, setting);
}

// float
void printParameter(const std::string& name, const float value, int len, bool setting) {
    printParameter(name, std::to_string(value), len, setting);
}

// double
void printParameter(const std::string& name, const double value, int len, bool setting) {
    printParameter(name, std::to_string(value), len, setting);
}
*/

// bool
//void printParameter(const std::string& name, const bool value, int len, bool setting) {
//    printParameter(name, (value ? "true" : "false") /*to_string(value)*/, len, setting);
//}

////

void printProcessing() {
    syslib::printMessage(MESSAGE_PROCESSING);
}

void printStartTask(const std::string &task) {
    // format message
    syslib::printfMessage(MESSAGE_START_TASK, task);
}

void printTotalTimeStr(const std::string &timeStr) {
    // format message
    syslib::printfMessage(MESSAGE_TOTAL_TIME, timeStr);
}

void printTotalTime(long time) {
    printTotalTimeStr(formatTime(time));
}
    
void printTotalTimeInMilliseconds(long time) {
    printTotalTimeStr(formatTimeInMilliseconds(time));
}
    
void printTotalTimeInSeconds(long time) {
    printTotalTimeStr(formatTimeInSeconds(time));
}
    
// format

std::string formatParameterName(const std::string& name, int len) {
    // -<name>......
    return "-" + (len < 1 ? name : strlib::rpad(name, len, "."));
}

std::string formatParameterValue(const std::string& value, bool setting) {
    // <value>
    // not setting
    return (setting ? value : MESSAGE_NOT_SETTING);
}

std::string formatTimeInMilliseconds(long time) {
    return std::to_string(time) + " ms";
}

std::string formatTimeInSeconds(long time) {
    return std::to_string(time / 1000.0) + " s";
}

std::string formatTime(long time) {
    if (time > 1000) {
        return formatTimeInSeconds(time);        // in seconds
    } else {
        return formatTimeInMilliseconds(time);   // in milliseconds
    }
}

////

int maxParameterLen(const std::string array[], int len) {
    return strlib::maxLen(array, len) + 2; // '<name>..'
}

int maxParameterLen(Parameters* parameters) {

    if (parameters == nullptr) {
        return 0;
    }

    int result = 0;
    int len = 0;
    int count = parameters->getParameterCount();

    Parameter* parameter = nullptr;
    std::string name;

    for (int i = 0; i < count; i++) {
        parameter = parameters->getParameter(i);
        name = parameter->getName();
        len = name.length();
        if (len > result) {
            result = len;
        }
    }

    return result + 2; // '<name>..'
}

// fs
std::string getOptionalPath(const std::string& path1, const std::string& path2) {
    return fslib::getOptionalPath(path1, path2);
}

// fs
std::string generateFileName(std::string& dir, std::string& filePrefix, int number, std::string& fileExt) {
  return fslib::generateFileName(dir, filePrefix, number, fileExt);
}

// fs
std::string generateFileName(std::string& filePrefix, int number, std::string& fileExt) {
  return fslib::generateFileName(filePrefix, number, fileExt);
}

File: src/main/java/plazma/lib/task/TaskHelper.java
/*
 * Copyright (C) 2012-2023 Oleh Hapon ohapon@users.sourceforge.net
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
 * 
 * Oleh Hapon
 * Kyiv, UKRAINE
 * ohapon@users.sourceforge.net
 */

package plazma.lib.task;

import static plazma.lib.sys.SysLib.*;

import plazma.lib.fmt.FmtLib;
import plazma.lib.fs.FSLib;
import plazma.lib.str.StrLib;
import plazma.lib.sys.Parameter;
import plazma.lib.sys.Parameters;

public class TaskHelper {

    public static final String MESSAGE_START_TASK = "Start task: '%s'";
    public static final String MESSAGE_PROCESSING = "Processing...";
    public static final String MESSAGE_TOTAL_TIME = "Total time: %s";

    public static final String MESSAGE_NOT_SETTING = "<not setting>";
    public static final String MESSAGE_BY_DEFAULT = "by default";
    
    
    public static boolean isEmpty(String str) {
        return StrLib.isEmpty(str);
    }

    public static boolean hasValue(String str) {
        return !isEmpty(str);
    }

    /**
     * Parse argument array and return parameter map
     * 
     * @param args
     * @return
     */

    ////
    
    public static void printParameter(Parameter parameter, int len) {
        if (parameter == null) {
            return;            
        }
        String name = parameter.getName();
        String value = parameter.getValue();
        String dataType = parameter.getDataType();
        boolean setting = parameter.hasValue();
        
        String displayName = formatName(name, len);
        String displayValue = formatValue(value, setting);
        
        //if ("Date".equalsIgnoreCase(dataType)) {
        //    displayValue = format
        //}        

        println(" " + displayName + ": " + displayValue);
    }

    
    // string
    public static void printParameter(String name, String value, int len, boolean setting) {
        println(" " + formatName(name, len) + ": " + formatValue(value, setting));
    }

    /*
    public static void printParameter(String name, String value) {
        printParameter(name, value, 0, true);
    }

    public static void printParameter(String name, String value, boolean setting) {
        printParameter(name, value, 0, setting);
    }

    // string
    public static void printParameter(String name, String value, int len) {
        printParameter(name, value, len, true);
    }

    // int
    public static void printParameter(String name, int value, int len, boolean setting) {
        printParameter(name, toString(value), len, setting);
    }

    // int
    public static void printParameter(String name, int value, int len) {
        printParameter(name, value, len, true);
    }

    // long
    public static void printParameter(String name, long value, int len, boolean setting) {
        printParameter(name, toString(value), len, setting);
    }

    // long
    public static void printParameter(String name, long value, int len) {
        printParameter(name, value, len, true);
    }

    // float
    public static void printParameter(String name, float value, int len, boolean setting) {
        printParameter(name, toString(value), len, setting);
    }

    // float
    public static void printParameter(String name, float value, int len) {
        printParameter(name, value, len, true);
    }

    // double
    public static void printParameter(String name, double value, int len, boolean setting) {
        printParameter(name, toString(value), len, setting);
    }

    // double
    public static void printParameter(String name, double value, int len) {
        printParameter(name, value, len, true);
    }

    // boolean
    public static void printParameter(String name, boolean value, int len, boolean setting) {
        printParameter(name, toString(value), len, setting);
    }

    // boolean
    public static void printParameter(String name, boolean value, int len) {
        printParameter(name, value, len, true);
    }
    */

    ////

    // int
    public static String toString(int value) {
        return "" + value;
    }

    // long
    public static String toString(long value) {
        return "" + value;
    }

    // float
    public static String toString(float value) {
        return "" + value;
    }

    // double
    public static String toString(double value) {
        return "" + value;
    }

    // boolean
    public static String toString(boolean value) {
        return (value ? "true" : "false");
    }

    ////

    public static void printStartTask(String task) {
        println(format(MESSAGE_START_TASK, task));
    }

    public static void printProcessing() {
        println(MESSAGE_PROCESSING);
    }

    public static void printTotalTime(long time) {
        println(format(MESSAGE_TOTAL_TIME, formatTime(time)));
    }

    // format

    public static final String format(String message, Object... args) {
        return FmtLib.format(message, args);
    }

    public static String formatName(String name, int len) {
        // -<name>......
        return "-" + (len < 1 ? name : StrLib.rpad(name, len, "."));
    }

    public static String formatValue(String value, boolean setting) {
        // <value>
        // not setting
        return (setting ? value : MESSAGE_NOT_SETTING);
    }

    // string
    public static String formatByDefault(String value, boolean def) {
        // <value>
        // <value> by default
        return "" + value + (def ? (" " + MESSAGE_BY_DEFAULT) : "");
    }

    // int
    public static String formatByDefault(int value, boolean def) {
        return formatByDefault(toString(value), def);
    }

    // long
    public static String formatByDefault(long value, boolean def) {
        return formatByDefault(toString(value), def);
    }

    // float
    public static String formatByDefault(float value, boolean def) {
        return formatByDefault(toString(value), def);
    }

    // double
    public static String formatByDefault(double value, boolean def) {
        return formatByDefault(toString(value), def);
    }

    // ms, s
    public static String formatTime(long time) {
        if (time > 1000) {
            return formatTime(time / 1000.0f, "s");
        }
        return formatTime(time, "ms");
    }

    public static String formatTime(long time, String unit) {
        return "" + time + " " + unit;
    }

    public static String formatTime(float time, String unit) {
        return "" + time + " " + unit;
    }

    ////

    public static int maxParameterLen(String[] array) {
        return StrLib.maxLen(array) + 2; // '<name>..'
    }
    
    public static int maxParameterLen(Parameters parameters) {

        if (parameters == null) {
            return 0;
        }

        int result = 0;
        int len = 0;
        int count = parameters.getParameterCount();

        Parameter parameter = null;
        String name;

        for (int i = 0; i < count; i++) {
            parameter = parameters.getParameter(i);
            name = parameter.getName();
            len = name.length();
            if (len > result) {
                result = len;
            }
        }

        return result + 2; // '<name>..'
    }
  

    ////

    public static String getOptionalPath(String dir, String filePrefix) {
        return FSLib.getOptionalPath(dir, filePrefix);
    }

    public static String generateFileName(String dir, String filePrefix, int number, String fileExt) {
        return FSLib.generateFileName(dir, filePrefix, number, fileExt);
    }

    public static String generateFileName(String filePrefix, int number, String fileExt) {
        return FSLib.generateFileName(filePrefix, number, fileExt);
    }
    
    //////////////////////////////////////////////////////////////////////////////////
    
    /*
    
    //// print

    protected void printHeader(String header) {
        println();
        if (header != null) {
            printLine();
            println(header);
        }
        printLine();
    }

    protected void printLine() {
        println("=============================================================================");
    }

    ////

    protected String format(Object obj, int len) {
        String str = obj == null ? "" : obj.toString();
        return StrLib.formatString(str, len);
    }
    */


}
